File: src/Servidor/Transaccional/AccesoDatos/Dao/NHibernate/DaoBaseNHibernate.cs
using System;
using System.Collections.Generic;
using NHibernate;
using NHibernate.Criterion;
using NLog;
using Trascend.Bolet.AccesoDatos.Contrato;
using Trascend.Bolet.ObjetosComunes.Entidades;

namespace Trascend.Bolet.AccesoDatos.Dao.NHibernate
{
    public abstract class DaoBaseNHibernate<T, Id> : IDaoBase<T, Id>
    {

        private static Logger logger = LogManager.GetCurrentClassLogger();
        private ISession _session;
        //private NHibernateHelper _helper;

        //public DaoBaseNHibernate()
        //{
        //    _helper = NHibernateHelper.ObtenerInstanciaNHibernateHelper();
        //}


        /// <summary>
        /// Propiedad para obtener la sesion
        /// </summary>
        public ISession Session
        {
            get
            {
                if (null == _session)
                    _session = NHibernateHelper.OpenSession();
                return _session;
            }

            set { _session = value; }
        }


        /// <summary>
        /// Método que obtiene todos los registros de la entidad
        /// </summary>
        /// <returns>Lista con todos los elementos de la entidad</returns>
        public IList<T> ObtenerTodos()
        {
            IList<T> listaEntidad;

            try
            {
                listaEntidad = Session.CreateCriteria(typeof(T)).AddOrder(Order.Asc("Id")).List<T>();
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExConsultarTodos);
            }
            finally
            {
                Session.Close();
            }

            return listaEntidad;
        }


        /// <summary>
        /// Método que obtiene todos los registros de la entidad ordenada en el orden especificado
        /// </summary>
        /// <param name="parametro">Columna por la que se desea ordenar</param>
        /// <param name="orden">"Asc" para ascendiente, "Desc" para descendiente</param>
        /// <returns>La Lista ordenada en lo especificado</returns>
        public IList<T> ObtenerTodos(string parametro, string orden)
        {
            IList<T> listaEntidad;

            try
            {
                if (orden.Equals("Asc"))
                    listaEntidad = Session.CreateCriteria(typeof(T)).AddOrder(Order.Asc(parametro)).List<T>();
                else
                    listaEntidad = Session.CreateCriteria(typeof(T)).AddOrder(Order.Desc(parametro)).List<T>();

            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExConsultarTodos);
            }
            finally
            {
                Session.Close();
            }

            return listaEntidad;
        }


        /// <summary>
        /// Método que obtiene el elemento de la entidad por su id
        /// </summary>
        /// <param name="id">Id de la entidad a requerida</param>
        /// <returns>Entidad requerida</returns>
        public T ObtenerPorId(Id id)
        {
            T entidad;

            try
            {
                entidad = Session.Get<T>(id);
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExObtenerPorId);
            }
            finally
            {
                Session.Close();
            }

            return entidad;
        }


        /// <summary>
        /// Método que verifica si existe una entidad
        /// </summary>
        /// <param name="entidad">Entidad a verificar</param>
        /// <returns>True si existe, false en caso contrario</returns>
        public bool VerificarExistencia(Id id)
        {
            bool existe = false;
            T entidad;

            try
            {
                entidad = Session.Get<T>(id);

                if (entidad != null)
                    existe = true;
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExObtenerPorId);
            }
            finally
            {
                Session.Close();
            }

            return existe;
        }


        /// <summary>
        /// Método que obtiene el elemento de la entidad por su id y 
        /// lo bloquea hasta que sea actualizado
        /// </summary>
        /// <param name="id">Id de la entidad a requerida</param>
        /// <returns>Entidad requerida</returns>
        public T ObtenerPorIdYBloquear(Id id)
        {
            T entidad;

            try
            {
                entidad = Session.Get<T>(id, LockMode.Upgrade);
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExObtenerPorId);
            }
            finally
            {
                Session.Close();
            }

            return entidad;
        }


        /// <summary>
        /// Método que inserta o modifica una entidad
        /// </summary>
        /// <param name="entidad">Entidad a modificar</param>
        /// <returns>True si fue éxitoso la inserción o modificación, en caso contrario False</returns>
        public bool InsertarOModificar(T entidad)
        {
            bool exitoso;

            try
            {
                ITransaction transaccion = Session.BeginTransaction();
                Session.SaveOrUpdate(entidad);
                transaccion.Commit();
                exitoso = transaccion.WasCommitted;
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                //throw new ApplicationException(Recursos.Errores.ExInsertarOModificar);
                throw new ApplicationException(Recursos.Errores.ExInsertarOModificar + ": " + ex.InnerException.Message);
            }
            finally
            {
                Session.Close();
            }

            return exitoso;
        }


        /// <summary>
        /// Elimina la Entidad que se envio
        /// </summary>
        /// <param name="entidad">Elimina la entidad</param>
        /// <returns>True si se elimino, de lo contrario false</returns>
        public bool Eliminar(T entidad)
        {
            bool exitoso;

            try
            {
                ITransaction transaccion = Session.BeginTransaction();
                Session.Delete(entidad);
                transaccion.Commit();
                exitoso = transaccion.WasCommitted;
            }
            catch (Exception ex)
            {
                logger.Error(ex.Message);
                throw new ApplicationException(Recursos.Errores.ExEliminando);
            }
            finally
            {
                Session.Close();
            }

            return exitoso;
        }



        ///// <summary>
        ///// Método que cierra la sesion de la BD
        ///// </summary>
        //public void CerrarSesion()
        //{
        //    try
        //    {
        //        _helper.GetSession().Flush();
        //        _helper.GetSession().Clear();
        //    }
        //    catch (Exception ex)
        //    {
        //        string a = ex.Message;
        //        throw;
        //    }
        //}
    }
}

File: trunk/src/Servidor/Transaccional/AccesoDatos_Fac/Dao/NHibernate/DaoBaseNHibernate.vb
Imports System.Collections.Generic
Imports NHibernate
Imports NHibernate.Criterion
Imports NLog
Imports Diginsoft.Bolet.AccesoDatos.Contrato
Imports Diginsoft.Bolet.ObjetosComunes.Entidades

Namespace Dao.NHibernate
    Public MustInherit Class DaoBaseNHibernate(Of T, Id)
        Implements IDaoBase(Of T, Id)

        Private Shared logger As Logger = LogManager.GetCurrentClassLogger()
        Private _session As ISession

        ''' <summary>
        ''' Propiedad para obtener la sesion
        ''' </summary>
        Public Property Session() As ISession
            Get
                If _session Is Nothing Then
                    _session = NHibernateHelper.OpenSession()
                End If
                Return _session
            End Get

            Set(ByVal value As ISession)
                _session = value
            End Set
        End Property



        Public Function Eliminar(ByVal entidad As T) As Boolean Implements Contrato.IDaoBase(Of T, Id).Eliminar
            Dim exitoso As Boolean

            Try
                Dim transaccion As ITransaction = Session.BeginTransaction()
                Session.Delete(entidad)
                transaccion.Commit()
                exitoso = transaccion.WasCommitted
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExEliminando)
            Finally
                Session.Close()
            End Try

            Return exitoso
        End Function

        Public Function InsertarOModificar(ByVal entidad As T) As Boolean Implements Contrato.IDaoBase(Of T, Id).InsertarOModificar
            Dim exitoso As Boolean

            Try
                Dim transaccion As ITransaction = Session.BeginTransaction()
                Session.SaveOrUpdate(entidad)
                transaccion.Commit()
                exitoso = transaccion.WasCommitted
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExInsertarOModificar)
            Finally
                Session.Close()
            End Try

            Return exitoso
        End Function

        Public Function ObtenerPorId(ByVal id As Id) As T Implements Contrato.IDaoBase(Of T, Id).ObtenerPorId
            Dim entidad As T

            Try
                entidad = Session.[Get](Of T)(id)
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExObtenerPorId)
            Finally
                Session.Close()
            End Try

            Return entidad
        End Function

        Public Function ObtenerPorIdYBloquear(ByVal id As Id) As T Implements Contrato.IDaoBase(Of T, Id).ObtenerPorIdYBloquear
            Dim entidad As T

            Try
                entidad = Session.[Get](Of T)(id, LockMode.Upgrade)
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExObtenerPorId)
            Finally
                Session.Close()
            End Try

            Return entidad
        End Function

        Public Function ObtenerTodos() As System.Collections.Generic.IList(Of T) Implements Contrato.IDaoBase(Of T, Id).ObtenerTodos
            Dim listaEntidad As IList(Of T)

            Try
                listaEntidad = Session.CreateCriteria(GetType(T)).AddOrder(Order.Asc("Id")).List(Of T)()
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExConsultarTodos)
            Finally
                Session.Close()
            End Try

            Return listaEntidad
        End Function

        Public Function VerificarExistencia(ByVal id As Id) As Boolean Implements Contrato.IDaoBase(Of T, Id).VerificarExistencia
            Dim existe As Boolean = False
            Dim entidad As T

            Try
                entidad = Session.[Get](Of T)(id)

                If entidad IsNot Nothing Then
                    existe = True
                End If
            Catch ex As Exception
                logger.[Error](ex.Message)
                Throw New ApplicationException(Recursos.Errores.ExObtenerPorId)
            Finally
                Session.Close()
            End Try

            Return existe
        End Function
    End Class
End Namespace